airbyte-integrations/connectors/source-linkedin-ads/source_linkedin_ads/utils.py

```diff
@@ -0,0 +1,183 @@
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+#
+
+from typing import Dict, Iterable, List, Mapping
+
+import pendulum as pdm
+
+
+def make_slice(record: Dict, key_value_map: Dict) -> Dict:
+    """
+    Outputs the Dict with key:value slices for the stream.
+    EXAMPLE:
+        in: records = [{dict}, {dict}, ...],
+            key_value_map = {<slice_key_name>: <key inside record>}
+        {
+            <slice_key_name> : records.<key inside record>.value,
+            ....
+        }
+    """
+    result = {}
+    for key in key_value_map:
+        value = record[key_value_map.get(key, None)]
+        if value:
+            result.update(**{key: value})
+    return result
+
+
+def transform_data(records: List) -> Iterable[Mapping]:
+    """
+    We need to transform the nested complex data structures into simple key:value pair,
+    to be properly normalised in the destination.
+
+    The cursor_field has the nested structure as:
+    :: EXAMPLE `dateRange` structure in Analytics streams:
+        {
+            "dateRange": {
+                "start": {"month": 8,"day": 13,"year": 2021},
+                "end": {"month": 8,"day": 13,"year": 2021}
+            }
+        }
+
+    :: EXAMPLE `changeAuditStamps` structure:
+        {
+            "changeAuditStamps": {
+                "created": {"time": 1629581275000},
+                "lastModified": {"time": 1629664544760}
+            }
+        }
+
+    """
+
+    for record in records:
+
+        # Transform `changeAuditStamps`
+        if "changeAuditStamps" in record:
+            dict_key: str = "changeAuditStamps"
+            props: List = ["created", "lastModified"]
+            fields: List = ["time"]
+
+            target_dict: Dict = record.get(dict_key, None)
+            if target_dict:
+                for prop in props:
+                    # Update dict with flatten key:value
+                    for field in fields:
+                        record.update(**{prop: pdm.from_timestamp(target_dict.get(prop).get(field, None) / 1000).to_datetime_string()})
+                record.pop(dict_key)
+
+        # Transform `dateRange`
+        if "dateRange" in record:
+            dict_key: str = "dateRange"
+            props: List = ["start", "end"]
+            fields: List = ["year", "month", "day"]
+
+            target_dict: Dict = record.get(dict_key, None)
+            if target_dict:
+                for prop in props:
+                    # Update dict with flatten key:value
+                    for field in fields:
+                        record.update(**{f"{prop}.{field}": target_dict.get(prop).get(field, None)})
+                # We build `start_date` & `end_date` fields from nested structure.
+                record.update(
+                    **{
+                        "start_date": pdm.date(record["start.year"], record["start.month"], record["start.day"]).to_date_string(),
```

Comment:
seems like the code like "pdm.date(record["start.year"], record["start.month"], record["start.day"]).to_date_string()"  - is a candidate for function

Reply:
Done. Made `date_str_from_date_range()` method for this.